rllib/agents/trainer.py

```diff
@@ -601,20 +622,40 @@ def train(self) -> ResultDict:
         for _ in range(1 + MAX_WORKER_FAILURE_RETRIES):
             try:
                 result = Trainable.train(self)
-            except RayError as e:
-                if self.config["ignore_worker_failures"]:
-                    logger.exception(
-                        "Error in train call, attempting to recover")
-                    self._try_recover()
-                else:
-                    logger.info(
-                        "Worker crashed during call to train(). To attempt to "
-                        "continue training without the failed worker, set "
-                        "`'ignore_worker_failures': True`.")
-                    raise e
             except Exception as e:
-                time.sleep(0.5)  # allow logs messages to propagate
-                raise e
+                if issubclass(e, RayError):
+                    # do not retry in case of OOM errors
+                    if is_memory_error(e):
+                        logger.exception("Not attempting to recover from error in train call "
+                                         "since it was caused by an OOM error",
+                                         exc_info=e)
+                        time.sleep(0.5)  # allow logs messages to propagate
+                        raise e
+                    else:
+                        # always retry on NoSamplesAvailable as this is by definition
+                        # a retryable situation
+                        if isinstance(e, NoSamplesAvailable):
```

Comment:
Rather than catching it here and doing `_try_recover()`, we need to handle it in our bonsai code (I'll show you).